.github/workflows/ci.yml

```diff
@@ -15,85 +15,51 @@ on:
       - 'LICENCE'
       - 'CONTRIBUTION.md'
       - 'README.md'
+
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
+
 jobs:
+  load_matrix:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix_build: ${{ steps.set-matrix.outputs.matrix_build }}
+      matrix_test: ${{ steps.set-matrix.outputs.matrix_test }}
+
+    steps:
+      - uses: actions/checkout@v4
+      - id: set-matrix
+        shell: bash
+        run: |
+          echo "matrix_build=$(cat .github/workflows/matrix_build.json | jq -c .)" >> $GITHUB_OUTPUT
+          echo "matrix_test=$(cat .github/workflows/matrix_test.json | jq -c .)" >> $GITHUB_OUTPUT
+
   build:
+    needs: load_matrix
     strategy:
-      matrix:
-        os: [linux, macos, windows]
-        config: [debug, release]
-        compiler: [gcc, clang, cl]
-        platform: [x86_64, aarch64]
-        exclude:
-          # Default to x64, but aarch64 on osx
-          - { os: linux, platform: aarch64 }
-          - { os: windows, platform: aarch64 }
-          - { os: macos, platform: x86_64 }
-          # Unused compiler configs
-          - { os: linux, compiler: clang }
-          - { os: linux, compiler: cl }
-          - { os: windows, compiler: gcc }
-          - { os: windows, compiler: clang }
-          - { os: macos, compiler: gcc }
-          - { os: macos, compiler: cl }
-        include:
-          - { os: linux, runs-on: ubuntu-20.04 }
-          - { os: macos, runs-on: macos-latest }
-          - { os: windows, runs-on: windows-latest }
-          # Warnings are treated as errors by default.
-          # But we may want to disable it temporarily.
-          - { os: linux, warnings-as-errors: true }
-          - { os: macos, warnings-as-errors: true }
-          - { os: windows, warnings-as-errors: true }
-          # Set a test category depending on the config, smoke by default,
-          # quick or full conditionally otherwise
-          - test-category: smoke
-          - { os: windows, test-category: quick }
-          - { config: release, test-category: full }
-          # default not full gpu tests
-          - full-gpu-tests: false
-          # The runners don't have a GPU by default except for the self-hosted ones
-          - has-gpu: false
-          # Self-hosted aarch64 build
-          - os: linux
-            config: release
-            compiler: gcc
-            platform: aarch64
-            test-category: smoke
-            full-gpu-tests: false
-            runs-on: [self-hosted, Linux, ARM64]
-            has-gpu: true
-          # Self-hosted full gpu build
-          - os: windows
-            config: release
-            compiler: cl
-            platform: x86_64
-            test-category: full
-            full-gpu-tests: true
-            runs-on: [Windows, self-hosted]
-            has-gpu: true
       fail-fast: false
+      matrix: ${{ fromJson(needs.load_matrix.outputs.matrix_build) }}
```

Comment:
I think it's probably neater to have this inline, because then we can at least have comments. Is it possible to do the GITHUB_OUTPUT thing without having that file separate?

Reply:
As I explained on the description of the PR, I am planning to delete the new files, "matrix_build.json" and "matrix_test.json".
I will inline them in the CI.yml.
Because json file doesn't allow us to leave any comments, I think it is better to keep them in YAML file as it has been.